modules/_labs/pagination/react/lib/Pages.tsx

```diff
@@ -0,0 +1,138 @@
+  display: 'inline-flex',
+  alignItems: 'center',
+  justifyContent: 'center',
+  fontSize: 13,
+  fontWeight: 700,
+  height: 32,
+  width: 32,
+  backgroundColor: canvas.colors.blueberry400,
+  color: canvas.colors.frenchVanilla100,
+  borderRadius: borderRadius.m,
+  ':focus': {outline: 'none'},
+});
+
+const NoPointerEvents = styled('div')({
+  pointerEvents: 'none',
+});
+
+const Ellipse = () => (
+  <NoPointerEvents>
+    <IconButton
+      key={'ellipse'}
+      aria-label={`Navigation Ellipse`}
+      variant={IconButton.Variant.Square}
+      size={IconButton.Size.Small}
+      tabIndex={-1}
+    >
+      ...
+    </IconButton>
+  </NoPointerEvents>
+);
+
+const Pages: React.FC<{
+  numPages: number;
+  currentPage: number;
+  clickHandler: (page: number) => void;
+}> = props => {
+  const {numPages, currentPage, clickHandler} = props;
+  const [width, setWidth] = React.useState(window.innerWidth);
+
+  React.useEffect(() => {
+    const handleWindowSizeChange = () => setWidth(window.innerWidth);
+    window.addEventListener('resize', handleWindowSizeChange);
+    return () => window.removeEventListener('resize', handleWindowSizeChange);
+  }, []);
+  const mobile = width < 500;
+
+  let pagesToDisplay = 5;
+  let start = 1;
+
+  if (mobile) {
+    start = currentPage;
+    if (currentPage >= numPages - 1) {
+      start = Math.max(numPages - 2, 1);
+      pagesToDisplay = 3;
+    } else {
+      pagesToDisplay = 1;
+    }
+  } else {
+    const midwayAndMorePages = currentPage > 2 && numPages > pagesToDisplay;
+    if (midwayAndMorePages) {
+      const oneOfLastThreePages = currentPage >= numPages - 3;
+      if (oneOfLastThreePages) {
+        start = numPages - 4;
+      } else {
+        start = currentPage - 2;
+      }
+    }
+  }
+
+  const end = Math.min(start + pagesToDisplay, numPages + 1);
+  const pages = _.range(start, end);
+
+  const PaginationButton: React.FC<{
+    page: number;
+  }> = props => (
+    <IconButton
+      data-testid={`paginationButton${props.page}`}
+      key={props.page}
+      aria-label={`Page ${props.page}`}
+      variant={IconButton.Variant.Square}
+      size={IconButton.Size.Small}
+      onClick={e => clickHandler(props.page)}
+    >
+      {props.page}
+    </IconButton>
+  );
+
+  return (
+    <>
+      {pages.map(page => {
+        const active = page === currentPage;
+        const more = page === end - 1 && page < numPages;
+        const lastPage = page === start && end === numPages + 1 && numPages > pagesToDisplay;
+
+        return (
+          <>
+            {lastPage && !mobile && (
+              <>
+                <PaginationButton page={1} />
+                <Ellipse />
+              </>
+            )}
+
+            {active ? (
+              <ActivePage
+                data-testid="paginationIconButtonActive"
+                key={page}
+                aria-current={true}
+                aria-label={`Page ${page}, Selected`}
+                autoFocus
```

Comment:
I noticed some strange behavior of this component and this is the cause. When I click on a page, it comes into focus (even though the focus styling is removed). But when I click on the "Next" or "Back" buttons, they visibly and actually lose focus because this component keeps taking it again.

All of the branching logic that's happening here causes the React reconciler to destroy and recreate DOM since we're using style components instead of `div` components with a `css` property that changes depending on the type. This reconciler action normally isn't too bad (other than increased GC and overall less performant), but it does get rid of DOM state like focus and selection.

This should be fixed before this component leaves Labs

Reply:
The reason I always give focus to the active page on render is because I was told by design that it should always hold focus. I agree it's strange, so I'm removing the autofocus logic.

Regarding your second point, I can refactor to avoid this.